EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -138,90 +158,123 @@ void setContactorsTask(void const* arg)
-                uint8_t charge = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
-                uint8_t discharge = !HAL_GPIO_ReadPin(DISCHARGE_SENSE_GPIO_Port, DISCHARGE_SENSE_Pin);
+                common = !HAL_GPIO_ReadPin(COMMON_SENSE_GPIO_Port, COMMON_SENSE_Pin);
+                charge = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
+                discharge = !HAL_GPIO_ReadPin(DISCHARGE_SENSE_GPIO_Port, DISCHARGE_SENSE_Pin);
+
+                uint8_t turningOnCharge = 0;
+                uint8_t turningOnDischarge = 0;
 
-                if (!common || !charge || !discharge)
+                if (orionStatus.shutOff)
                 {
-                    // If any of the contactors are not enabled
-                    state = FIRST_CHECK;
+                    state = SHUTDOWN;
+                    continue;
                 }
 
-                break;
-            }
-
-            case BLOCKED:
-                if (orionStatus.gpioOk && orionStatus.batteryVoltagesInRange)
+                if (orionStatus.allowCharge && orionStatus.chargeContactorOverriden && !charge)
                 {
-                    state = FIRST_CHECK;
+                    if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+                    {
+                        continue;
+                    }
+
+                    orionStatus.chargeContactorOverriden = 0;
+                    osMutexRelease(orionStatus.orionStatusMutex);
+
+                    turningOnCharge = 1;
+                    // Turn on charge Contactor and go back to recheck
+                    HAL_GPIO_WritePin(CHARGE_CONTACTOR_ENABLE_GPIO_Port, CHARGE_CONTACTOR_ENABLE_Pin, GPIO_PIN_SET);
+                    state = CHARGE_CONTACTOR_ENABLE_CHECK;
                 }
+                else if (orionStatus.allowDischarge && orionStatus.dischargeContactorOverriden && !discharge)
+                {
+                    if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+                    {
+                        continue;
+                    }
 
-                break;
+                    orionStatus.dischargeContactorOverriden = 0;
+                    osMutexRelease(orionStatus.orionStatusMutex);
 
-            default:
-                state = FIRST_CHECK;
-        }
-    }
-}
+                    turningOnDischarge = 1;
+                    // Turn on discharge contactor and go back to recheck
+                    HAL_GPIO_WritePin(DISCHARGE_CONTACTOR_ENABLE_GPIO_Port, DISCHARGE_CONTACTOR_ENABLE_Pin, GPIO_PIN_SET);
+                    state = DISCHARGE_CONTACTOR_ENABLE_CHECK;
+                }
 
-int updateContactorState(ContactorState newState, uint8_t error, Contactor contactor)
-{
-    if (osMutexWait(auxStatus.auxStatusMutex, 0) != osOK)
-    {
-        return 0;
-    }
+                if ((orionStatus.allowCharge && !(common && charge) && !turningOnCharge) ||
```

Comment:
Since common's so important, we should check if it has been disconnected first before doing anything else in the DONE state.